prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java

```
@@ -0,0 +1,267 @@
+        private final Consumer<List<Types.TimeSeries>> requestResult;
+        private final SinkCounters counters;
+
+        public ResponseCallback(
+                int timeSeriesCount,
+                long sampleCount,
+                SinkCounters counters,
+                Consumer<List<Types.TimeSeries>> requestResult) {
+            this.timeSeriesCount = timeSeriesCount;
+            this.sampleCount = sampleCount;
+            this.requestResult = requestResult;
+            this.counters = counters;
+        }
+
+        @Override
+        public void completed(SimpleHttpResponse response) {
+            if (RemoteWriteResponseClassifier.isSuccessResponse(response)) {
+                LOG.debug(
+                        "{},{} - successfully posted {} time-series, containing {} samples",
+                        response.getCode(),
+                        response.getReasonPhrase(),
+                        timeSeriesCount,
+                        sampleCount);
+                counters.inc(NUM_SAMPLES_OUT, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_OUT);
+            } else {
+                counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+
+                String responseBody = response.getBodyText();
+                if (RemoteWriteResponseClassifier.isNonRetriableErrorResponse(response)) {
+                    LOG.warn(
+                            "{},{} {} (discarded {} time-series, containing {} samples)",
+                            response.getCode(),
+                            response.getReasonPhrase(),
+                            responseBody,
+                            timeSeriesCount,
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_NON_RETRIABLE_DROPPED, sampleCount);
+                } else if (RemoteWriteResponseClassifier.isRetriableErrorResponse(response)) {
+                    LOG.warn(
```

We shouldn't be dropping records here on a retryable error.

IMO we should either requeue the records, or propagate up the exception to fail the job.

---

After offline discussion, we agreed that maybe we can set the HttpCllient to a lower maxRetryCount (by default), and then if the maxRetry is exceeded, we can throw the Exception to restart the Flink job

---

Modified the behaviour so that, by default, it will throw an exception and stop the job on any unrecoverable condition. Optionally, the user can decide not to block the job, but drop data and continue, selectively for three separate conditions:
1. Max retry exceeded
2. Non-retriable error reported
3. Http client I/O error
When the Http request is cancelled, an exception is always thrown (not configurable)

The default max retry is now 100, ~ 500 sec with the default backoff timing